### ISSUE TITLE
torch.distributed and RPC cannot both be initialized with the same host:port pair.

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#41614]
## 🐛 Bug
If users attempt to initialize `torch.distributed` (via `dist.init_process_group`) and `rpc` (via `rpc.init_rpc`), if they provide the same init option (such as a TCP host:port pair) to both of them, the initialization will fail.

## To Reproduce
Run the following script:
```
import torch
import torch.distributed as dist
import torch.distributed.rpc as rpc
import torch.multiprocessing as mp
import os
import tempfile

def run(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "29500"
    dist.init_process_group(
                backend="gloo",
                init_method="env://",
                rank=rank,
                world_size=world_size,
            )

    backend_opts = rpc.ProcessGroupRpcBackendOptions(init_method="tcp://localhost:29500")
    rpc.init_rpc(
        name=f"worker{rank}",
        rank=rank,
        world_size=world_size,
        rpc_backend_options=backend_opts,
    )

    print("Done with init")

    rpc.shutdown()


if __name__ == '__main__':
    ws = 2
    mp.spawn(run, args=(ws,), nprocs=2)

```

It fails with:

```
  File "/home/rvarm1/pytorch/torch/distributed/rpc/__init__.py", line 82, in init_rpc
    store, _, _ = next(rendezvous_iterator)
  File "/home/rvarm1/pytorch/torch/distributed/rendezvous.py", line 126, in _tcp_rendezvous_handler
    store = TCPStore(result.hostname, result.port, world_size, start_daemon, timeout)
RuntimeError: Address already in use
```

## Expected behavior

Ideally, the TCPStore can be reused and not reconstructed, so we won't run into the "Address in use" error? Although, open to discussion on whether we should support this or not. Maybe we can cache the `TCPStore` and return it if there is a TCPStore already created with that primary port? 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @jjlilley @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @xush6528

**Comment by agolynski** [comment]
One question here is how to manage the lifetime of such store (right now, rank 0 owns it and it must be kept alive, it also came up in https://github.com/pytorch/pytorch/issues/40633). Do we need to assume dependency between rpc system and process groups or we'd manage lifetime of store via distributed counter?

**Comment by rohan-varma** [comment]
@agolynski is there a pointer to see where `TCPStore` is actually destructed by the rank 0? I think the dist counter or what I mentioned in that issue could be a way to ensure the store isn't destroyed until all ranks agree. Although, error handling in this case might be tough.

**Comment by agolynski** [comment]
@rohan-varma: tcpStoreDaemon_ is deleted here:
https://github.com/pytorch/pytorch/blob/f003008d6ec46812407ff276da5241ae533eb5d1/torch/lib/c10d/TCPStore.cpp#L306
and TCPStore (TCPStore py object) is deleted once the corresponding rank is going away.

It's interesting that there is a dist barrier implemented here
https://github.com/pytorch/pytorch/blob/f003008d6ec46812407ff276da5241ae533eb5d1/torch/lib/c10d/TCPStore.cpp#L311
but it's only waiting until all workers are initialized. 

Maybe if isServer_ rank similarly waits for all workers to be destroyed before killing itself it is sufficient to deal with #40633 and to allow sharing as you've proposed?

**Comment by rohan-varma** [comment]
@agolynski Thanks for digging into that, that is interesting. It seems that we can just reuse that function `waitForWorkers` in shutdown as well (as you indicated), the comment in that function seems to indicate that is the intended use case as well.

**Comment by pritamdamania87** [comment]
Sharing the TCPStore might lead to unintended consequences where two different systems might overwrite the same set of keys if we're not careful with naming. One option would be to ensure we always use the PrefixStore with different prefixes to ensure we never end up overwriting the same keys?

In terms of lifetime management, can we just wrap the TCPStore in a shared_ptr and pass it to ProcessGroup and RPC?

**Comment by mrshenli** [comment]
Can we wrap the TCPStore into a PrefixStore, and let RPC use PrefixStore instead? Will that solve the key contention problem?

**Comment by mrshenli** [comment]
@froody and @blefaudeux brought up a good point that, it might be too much hurdle for users to initialize twice (c10d and RPC), it might be easier to use if we support initializing an RPC gang using an initialized ProcessGroup. 

I recall we discussed the possibility to have a new `init_rpc` API that takes a Store instead of `rank`, `world_size`, `init_method` etc. Can this be one step further for solving this issue? E.g., we let `ProcessGroup` expose an API to return its store, and then init RPC using that Store. 

There are different options to implement this behavior.

Option 1: As the c10d gang is stable with fixed ranks, the derived RPC gang can also stay that way and share the same rank/id with the ProcessGroup instance. 
Option 2: Let the RPC stay decoupled from the c10d ProcessGroup, and still allows dynamic join/leave. In this case, we cannot match the rank/id between ProcessGroup and RPC agents any more.

**Comment by mrshenli** [comment]
self-assigned to reserve for bootcamp.

**Comment by rohan-varma** [comment]
A similar issue (TCPStore address in use) prevents initializing, shutting down, and re-initializing RPC when using tensorpipe backend with TCP init: https://github.com/pytorch/pytorch/issues/46491. See repro in https://github.com/pytorch/pytorch/pull/48001.

**Comment by lw** [comment]
> should we bring down this TCP store when calling rpc.shutdown()?

Is this not what is happening? Calling rpc.shutdown() should destroy the agent and thus in turn destroy the store, since it's stored on the agent. Or did I miss anything? Is someone else keeping the store alive? Is there perhaps a reference cycle that causes us to leak the store?

**Comment by rohan-varma** [comment]
> > should we bring down this TCP store when calling rpc.shutdown()?
> 
> Is this not what is happening? Calling rpc.shutdown() should destroy the agent and thus in turn destroy the store, since it's stored on the agent. Or did I miss anything? Is someone else keeping the store alive? Is there perhaps a reference cycle that causes us to leak the store?

I guess this is not happening, as per the repro in #48001, we call rpc.shutdown() but when we attempt to re-init RPC, there is an Address already in use error indicating that the store has not been properly destroyed.

**Comment by cbalioglu** [comment]
Please see the recently merged #58331 which introduces the "multi-tenant" future in `TCPStore` to address this specific problem.
